core/src/main/java/io/questdb/std/fastdouble/FastDoubleMath.java

```diff
@@ -0,0 +1,1103 @@
+            // convert the integer into a double. This is lossless since
+            // 0 <= i <= 2^53 - 1.
+            double d = (double) significand;
+            //
+            // The general idea is as follows.
+            // If 0 <= s < 2^53 and if 10^0 <= p <= 10^22 then
+            // 1) Both s and p can be represented exactly as 64-bit floating-point values
+            // 2) Because s and p can be represented exactly as floating-point values,
+            // then s * p and s / p will produce correctly rounded values.
+            //
+            if (power < 0) {
+                d = d / DOUBLE_POWERS_OF_TEN[-power];
+            } else {
+                d = d * DOUBLE_POWERS_OF_TEN[power];
+            }
+            return (isNegative) ? -d : d;
+        }
+
+
+        // The fast path has now failed, so we are falling back on the slower path.
+
+        // We are going to need to do some 64-bit arithmetic to get a more precise product.
+        // We use a table lookup approach.
+        // It is safe because
+        // power >= DOUBLE_MIN_EXPONENT_POWER_OF_TEN
+        // and power <= DOUBLE_MAX_EXPONENT_POWER_OF_TEN
+        // We recover the mantissa of the power, it has a leading 1. It is always
+        // rounded down.
+        long factorMantissa = MANTISSA_64[power - DOUBLE_MIN_EXPONENT_POWER_OF_TEN];
+
+
+        // The exponent is 1023 + 64 + power + floor(log(5**power)/log(2)).
+        //
+        // 1023 is the exponent bias.
+        // The 64 comes from the fact that we use a 64-bit word.
+        //
+        // Computing floor(log(5**power)/log(2)) could be
+        // slow. Instead, we use a fast function.
+        //
+        // For power in (-400,350), we have that
+        // (((152170 + 65536) * power ) >> 16);
+        // is equal to
+        //  floor(log(5**power)/log(2)) + power when power >= 0,
+        // and it is equal to
+        //  ceil(log(5**-power)/log(2)) + power when power < 0
+        //
+        //
+        // The 65536 is (1<<16) and corresponds to
+        // (65536 * power) >> 16 ---> power
+        //
+        // ((152170 * power ) >> 16) is equal to
+        // floor(log(5**power)/log(2))
+        //
+        // Note that this is not magic: 152170/(1<<16) is
+        // approximately equal to log(5)/log(2).
+        // The 1<<16 value is a power of two; we could use a
+        // larger power of 2 if we wanted to.
+        //
+        long exponent = (((152170L + 65536L) * power) >> 16) + DOUBLE_EXPONENT_BIAS + 64;
+        // We want the most significant bit of digits to be 1. Shift if needed.
+        int lz = Long.numberOfLeadingZeros(significand);
+        significand <<= lz;
+        // We want the most significant 64 bits of the product. We know
+        // this will be non-zero because the most significant bit of digits is
+        // 1.
+        //before Java 18
```

Comment:
nit: why exactly Java 18?

Reply:
this is not my comment :)

Reply:
Just noticed that it originates from `fullMultiplication`. Still, it makes little sense as project Valhalla is not available in Java 18 to my best knowledge.

Reply:
done in https://github.com/questdb/questdb/pull/2527/commits/c84fbc9604fa6791384002d3b9a1b38b41759d9d